### PR TITLE
Improve typing of distributed.comp_modules.utils.all_gather

### LINKED ISSUE #1368: Improve typing of distributed.comp_modules.utils.all_gather
## 🚀 Feature

As a followup to the issue #1344 and comment https://github.com/pytorch/ignite/pull/1355#discussion_r499854850 it is needed to rework the typing of the `utils.all_gather` and the different realizations serial, native, hvd, xla.

### PR DESCRIPTION
**Pull request by gruebel**
Fixes #1368  

Description:
As suggested I added as return type `List[Number]` and adjusted the code to also reflect it.

### CODE REVIEW

File: ignite/distributed/comp_models/base.py [hunk 1]
```diff
@@ -271,8 +276,10 @@ def spawn(*args: Any, **kwargs: Any) -> None:
     def all_reduce(self, tensor: Union[torch.Tensor, Number], op: str = "sum") -> Union[torch.Tensor, Number]:
         return tensor
 
-    def all_gather(self, tensor: Union[torch.Tensor, Number]) -> Union[torch.Tensor, Number]:  # type: ignore
-        return tensor
+    def all_gather(
+        self, tensor: Union[torch.Tensor, Number, str]
+    ) -> Union[torch.Tensor, Number, List[Number], List[str]]:
+        return cast(Union[torch.Tensor, Number, List[Number], List[str]], tensor)
```
Comment (vfdev-5) [inline ignite/distributed/comp_models/base.py:282]: Let's implement it as
```python
def all_gather(self, tensor: Union[torch.Tensor, Number, str]) -> Union[torch.Tensor, List[Number], List[str]]:
    if isinstance(tensor, torch.Tensor):
        return tensor
    return cast(Union[List[Number], List[str]], [tensor, ])
```

And a test may be required to make codecov and myself happy :)

**Review by vfdev-5 — COMMENTED** [review]
@gruebel thanks for the PR !
As we modified the code, some tests should be updated (by fixing for current input and adding another for list output).
https://github.com/pytorch/ignite/pull/1370/checks?check_run_id=1217369119#step:5:1536

I left another comment for serial model's all gather implementation.

**Comment by vfdev-5** [comment]
@gruebel thanks for the update! Seems like there is something missing for XLA. I'll check it as I have more time this coming week.

**Review by vfdev-5 — APPROVED** [review]
Thanks @gruebel ! LGTM

**Comment by gruebel** [comment]
Do you know why the builds are so unstable lately. Quite often there are Runtime errors 😞

**Comment by vfdev-5** [comment]
@gruebel these fails are related to distributed computations tests. Sometimes it happens that a necessary port is busy and it fail everyone. We tried to understand why this happen but this is too flaky...

**Comment by gruebel** [comment]
I see, thanks for the info 😄

**Pull request merged**